evm/src/permutation.rs

```diff
@@ -68,6 +68,32 @@ impl<F: Field> GrandProductChallenge<F> {
     }
 }
 
+impl GrandProductChallenge<Target> {
+    pub(crate) fn combine_circuit<F: RichField + Extendable<D>, const D: usize>(
+        &self,
+        builder: &mut CircuitBuilder<F, D>,
+        terms: &[ExtensionTarget<D>],
+    ) -> ExtensionTarget<D> {
+        let zero = builder.zero();
+        let mut factor = ReducingFactorTarget::new(self.beta.to_ext_target(zero));
+        let reduced = factor.reduce(terms, builder);
+        builder.add_extension(reduced, self.gamma.to_ext_target(zero))
```

Comment:
nit: a tad simpler

```suggestion
        let reduced = reduce_with_powers_ext_circuit(terms, self.beta);
        let gamma = builder.convert_to_ext(self.gamma);
        builder.add_extension(reduced, gamma)
```

Reply:
Or maybe better to stick with `ReducingFactorTarget` if you take the suggestion below, so that the two functions can be similar.